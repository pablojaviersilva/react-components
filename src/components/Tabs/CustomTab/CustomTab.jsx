import React from "react";
import { Tab } from "react-tabs";
import PropTypes from "prop-types";
import classNames from "classnames/bind";
import styles from "./CustomTab.module.scss";

function CustomTab({ "aria-selected": ariaSelected, isPrevious, onClick, children, className, ...otherProps }) {
    const cx = classNames.bind(styles);
    return (
        <Tab
            aria-selected={ariaSelected}
            onClick={onClick}
            className={`${styles.CustomTab} ${cx(className)} ${isPrevious ? styles.previous : ""}`}
            data-ui="Tab"
            role="tab"
            {...otherProps}
        >
            {children}
        </Tab>
    );
}

export default CustomTab;

CustomTab.tabsRole = "Tab"; // Required field to use your custom Tab
CustomTab.propTypes = {
    /**
     * Content to be rendered
     */
    children: PropTypes.node.isRequired,
    "aria-selected": PropTypes.bool,
    isPrevious: PropTypes.bool,
    onClick: PropTypes.func,
    /**
     * Override or extend the styles applied to the component
     */
    className: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string), PropTypes.arrayOf(PropTypes.object)]),
};

CustomTab.defaultProps = {
    "aria-selected": false,
    className: "",
    isPrevious: false,
    onClick: () => {},
};
