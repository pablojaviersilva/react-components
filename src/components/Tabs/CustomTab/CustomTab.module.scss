:root{
  --Tab-color: #7D7D7D;
  --Tab-font-size: 12px;
  --Tab-selected-color: var(--Kenmei-Primary-Adele);
}
.CustomTab{
  color: var(--Tab-color);
  cursor: pointer;
  font-size: var(--Tab-font-size);
  font-weight: 400;
  overflow: hidden;
  position: relative;
  text-overflow: ellipsis;
  white-space: nowrap;
  outline: none;
  padding: 6px 0;
  display: flex;
  justify-content: center;
  background-color: var(--Kenmei-Light-Light3);
  &:last-child{
    margin-right: 0px!important;
  }

  &[aria-selected=false]{
    margin-right: 2px;
  }
  &[aria-selected=true]{
    color: var(--Tab-selected-color);
    border-bottom: 2px solid var(--Tab-selected-color);
    background-color: var(--Kenmei-Light-Light1);
  }
  &.previous{
    margin-right: 0px!important;
  }
}
