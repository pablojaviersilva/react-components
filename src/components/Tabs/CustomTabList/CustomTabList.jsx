import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames/bind";
import { TabList } from "react-tabs";

import styles from "./CustomTabList.module.scss";

function CustomTabList({ children, className, selectedIndex, setSelectedIndex }) {
    const cx = classNames.bind(styles);

    return (
        <div className={styles.tabwrapper}>
            <TabList className={`${styles.CustomTabList} ${cx(className)}`} data-ui="TabList">
                {React.Children.map(children, (child, index) =>
                    React.cloneElement(child, {
                        "aria-selected": selectedIndex === index,
                        "aria-posinset": index + 1,
                        "aria-setsize": React.Children.count(children),
                        onClick: () => setSelectedIndex(index),
                        isPrevious: selectedIndex === index + 1,
                    }),
                )}
            </TabList>
        </div>
    );
}

export default CustomTabList;

CustomTabList.tabsRole = "TabList"; // Required field to use your custom TabList
CustomTabList.propTypes = {
    /**
     * Content to be rendered
     */
    children: PropTypes.node.isRequired,
    selectedIndex: PropTypes.number,
    setSelectedIndex: PropTypes.func,
    /**
     * Override or extend the styles applied to the component
     */
    className: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string), PropTypes.arrayOf(PropTypes.object)]),
};

CustomTabList.defaultProps = {
    className: "",
    selectedIndex: 0,
    setSelectedIndex: () => {},
};
