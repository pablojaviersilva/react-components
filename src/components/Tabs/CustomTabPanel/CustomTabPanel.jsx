import React from "react";
import PropTypes from "prop-types";
import { TabPanel } from "react-tabs";

import "./react-tabs.scss";

function CustomTabPanel({ children, ...otherProps }) {
    return (
        <TabPanel
            data-ui="TabPanel"
            {...Object.keys(otherProps).reduce((acc, key) => {
                // Filter out props that should not be passed to TabPanel
                if (key !== "selectedIndex" && key !== "setSelectedIndex") {
                    acc[key] = otherProps[key];
                }
                return acc;
            }, {})}
        >
            {children}
        </TabPanel>
    );
}

export default CustomTabPanel;

CustomTabPanel.tabsRole = "TabPanel"; // Required field to use your custom TabPanel
CustomTabPanel.propTypes = {
    /**
     * Content to be rendered
     */
    children: PropTypes.node.isRequired,
};