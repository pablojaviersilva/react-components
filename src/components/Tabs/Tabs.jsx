/* eslint-disable max-len */
import React, { useState } from "react";
import PropTypes from "prop-types";
import { Tabs } from "react-tabs";
import classNames from "classnames/bind";

import styles from "./Tabs.module.scss";

function CustomTabs({ children, className }) {
    const cx = classNames.bind(styles);
    const [selectedIndex, setSelectedIndex] = useState(0); // Estado para la pestaÃ±a seleccionada

    const handleSelect = (index) => {
        setSelectedIndex(index);
    };

    return (
        <Tabs className={`${styles.Tabs} ${cx(className)}`} data-ui="Tabs" onSelect={handleSelect} selectedIndex={selectedIndex}>
            {React.Children.map(children, (child) => React.cloneElement(child, { selectedIndex, setSelectedIndex }))}
        </Tabs>
    );
}

export default CustomTabs;

CustomTabs.propTypes = {
    /**
     * Content to be rendered
     */
    children: PropTypes.node.isRequired,
    /**
     * Override or extend the styles applied to the component
     */
    className: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string), PropTypes.arrayOf(PropTypes.object)]),
};

CustomTabs.defaultProps = {
    className: "",
};
