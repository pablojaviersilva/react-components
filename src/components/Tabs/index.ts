export { default as Tabs } from "./Tabs";
export { default as Tab } from "./CustomTab/CustomTab";
export { default as TabList } from "./CustomTabList/CustomTabList";
export { default as TabPanel } from "./CustomTabPanel/CustomTabPanel";
