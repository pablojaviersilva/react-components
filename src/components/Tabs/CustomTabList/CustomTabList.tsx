// CustomTabList.tsx
import React, { ReactNode, FC } from "react";
import { TabWrapper, StyledTabList } from "./CustomTabList.styles"; // Ajusta la ruta según sea necesario

interface CustomTabListProps {
    children: ReactNode;
    className?: string;
    style?: React.CSSProperties;
    selectedIndex: number;
    setSelectedIndex: (index: number) => void;
}

const CustomTabList: FC<CustomTabListProps> = ({ children, className = "", style, selectedIndex, setSelectedIndex }) => {
    return (
        <TabWrapper>
            <StyledTabList className={className} data-ui="TabList" style={style}>
                {React.Children.map(children, (child, index) =>
                    React.isValidElement(child)
                        ? React.cloneElement(child, {
                              "aria-selected": selectedIndex === index,
                              "aria-posinset": index + 1,
                              "aria-setsize": React.Children.count(children),
                              onClick: () => setSelectedIndex(index),
                              isPrevious: selectedIndex === index + 1,
                              // eslint-disable-next-line @typescript-eslint/no-explicit-any
                          } as any) // Usa `as any` para evitar problemas de tipos con atributos ARIA
                        : null,
                )}
            </StyledTabList>
            {/* Si necesitas la línea debajo de las pestañas, descomenta esto */}
            {/* <div className="tabline" /> */}
        </TabWrapper>
    );
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
(CustomTabList as any).tabsRole = "TabList"; // Required field to use your custom TabList

export default CustomTabList;
