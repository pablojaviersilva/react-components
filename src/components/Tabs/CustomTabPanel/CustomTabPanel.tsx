import React, { ReactNode, CSSProperties, HTMLProps } from "react";
import { TabPanelWrapper } from "./CustomTabPanel.styles";

// Define el tipo de las propiedades aceptadas por CustomTabPanel
interface CustomTabPanelProps extends Omit<HTMLProps<HTMLDivElement>, "style"> {
    children: ReactNode;
    style?: CSSProperties;
}

const CustomTabPanel: React.FC<CustomTabPanelProps> = ({ children, style, ...otherProps }) => {
    // Filtra las propiedades que no deben pasarse al componente TabPanel
    const filteredProps = Object.entries(otherProps).reduce(
        (acc, [key, value]) => {
            if (key !== "selectedIndex" && key !== "setSelectedIndex") {
                acc[key] = value;
            }
            return acc;
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        {} as Record<string, any>,
    );

    return (
        <TabPanelWrapper
            className={`react-tabs__tab-panel ${otherProps["aria-selected"] ? "react-tabs__tab-panel--selected" : ""}`}
            data-ui="TabPanel"
            style={style}
            {...filteredProps}
        >
            {children}
        </TabPanelWrapper>
    );
};

// Define el rol de tabs para el componente CustomTabPanel
// eslint-disable-next-line @typescript-eslint/no-explicit-any
(CustomTabPanel as any).tabsRole = "TabPanel";

export default CustomTabPanel;
