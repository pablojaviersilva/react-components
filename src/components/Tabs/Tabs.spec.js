import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom"; // Para aserciones adicionales como 'toBeInTheDocument'
import Tabs from "./Tabs";
import CustomTabList from "./CustomTabList";
import CustomTab from "./CustomTab";
import CustomTabPanel from "./CustomTabPanel";

describe("CustomTabs component", () => {
    test("renders without crashing and matches snapshot", () => {
        const { container } = render(
            <Tabs>
                <CustomTabList>
                    <CustomTab>Tab 1</CustomTab>
                    <CustomTab>Tab 2</CustomTab>
                </CustomTabList>
                <CustomTabPanel>Panel 1</CustomTabPanel>
                <CustomTabPanel>Panel 2</CustomTabPanel>
            </Tabs>,
        );
        expect(container).toMatchSnapshot();
    });

    test("displays correct tab content based on selectedIndex", () => {
        render(
            <Tabs>
                <CustomTabList>
                    <CustomTab>Tab 1</CustomTab>
                    <CustomTab>Tab 2</CustomTab>
                </CustomTabList>
                <CustomTabPanel>Panel 1</CustomTabPanel>
                <CustomTabPanel>Panel 2</CustomTabPanel>
            </Tabs>,
        );

        // Verifica que el panel correspondiente al índice seleccionado esté en el DOM
        expect(screen.getByText("Panel 1")).toBeInTheDocument();
        expect(screen.queryByText("Panel 2")).not.toBeInTheDocument();

        // Simula un clic en la segunda pestaña
        fireEvent.click(screen.getByText("Tab 2"));

        // Verifica que el segundo panel ahora esté visible
        expect(screen.getByText("Panel 2")).toBeInTheDocument();
        expect(screen.queryByText("Panel 1")).not.toBeInTheDocument();
    });

    test("calls handleSelect and updates selectedIndex when a tab is clicked", () => {
        const { getByText } = render(
            <Tabs>
                <CustomTabList>
                    <CustomTab>Tab 1</CustomTab>
                    <CustomTab>Tab 2</CustomTab>
                </CustomTabList>
                <CustomTabPanel>Panel 1</CustomTabPanel>
                <CustomTabPanel>Panel 2</CustomTabPanel>
            </Tabs>,
        );

        // Verificar que el índice de la pestaña seleccionada inicialmente es 0
        const tab1 = getByText("Tab 1");
        expect(tab1).toHaveAttribute("aria-selected", "true");

        // Clic en la segunda pestaña
        fireEvent.click(getByText("Tab 2"));

        // Verificar que el índice de la pestaña seleccionada ha cambiado
        const tab2 = getByText("Tab 2");
        expect(tab2).toHaveAttribute("aria-selected", "true");
        expect(tab1).toHaveAttribute("aria-selected", "false");
    });
});
