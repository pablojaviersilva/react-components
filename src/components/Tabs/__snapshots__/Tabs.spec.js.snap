// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTabs component renders without crashing and matches snapshot 1`] = `
<div>
  <div
    class="undefined "
    data-rttabs="true"
    data-ui="Tabs"
  >
    <div>
      <ul
        class="undefined "
        data-ui="TabList"
        role="tablist"
      >
        <li
          aria-controls="react-tabs-1"
          aria-disabled="false"
          aria-posinset="1"
          aria-selected="true"
          aria-setsize="2"
          class="undefined   react-tabs__tab--selected"
          data-rttab="true"
          data-ui="Tab"
          id="react-tabs-0"
          role="tab"
          tabindex="0"
        >
          Tab 1
        </li>
        <li
          aria-controls="react-tabs-3"
          aria-disabled="false"
          aria-posinset="2"
          aria-selected="false"
          aria-setsize="2"
          class="undefined  "
          data-rttab="true"
          data-ui="Tab"
          id="react-tabs-2"
          role="tab"
        >
          Tab 2
        </li>
      </ul>
    </div>
    <div
      aria-labelledby="react-tabs-0"
      class="react-tabs__tab-panel react-tabs__tab-panel--selected"
      data-ui="TabPanel"
      id="react-tabs-1"
      role="tabpanel"
    >
      Panel 1
    </div>
    <div
      aria-labelledby="react-tabs-2"
      class="react-tabs__tab-panel"
      data-ui="TabPanel"
      id="react-tabs-3"
      role="tabpanel"
    />
  </div>
</div>
`;
